tools/azure-sdk-tools/ci_tools/functions.py

```diff
@@ -625,3 +631,194 @@ def discover_prebuilt_package(dist_directory: str, setup_path: str, package_type
     if prebuilt_package is not None:
         packages.append(prebuilt_package)
     return packages
+
+
+def is_package_compatible(
+    package_name: str,
+    package_requirements: List[Requirement],
+    immutable_requirements: List[Requirement],
+    should_log: bool = True,
+) -> bool:
+    """
+    This function accepts a set of requirements for a package, and ensures that the package is compatible with the immutable_requirements.
+
+    It is factored this way because we retrieve requirements differently according to the source of the package.
+        If published, we can get requires() from PyPI
+        If locally built wheel, we can get requires() from the metadata of the package
+        If local relative requirement, we can get requires() from a ParsedSetup of the setup.py for th package
+
+    :param List[Requirement] package_requirements: The dependencies of a dev_requirement file. This is the set of requirements that we are checking compatibility for.
+    :param List[Requirement] immutable_requirements: A list of requirements that the other packages must be compatible with.
+    """
+
+    for immutable_requirement in immutable_requirements:
+        for package_requirement in package_requirements:
+            if package_requirement.key == immutable_requirement.key:
+                # if the dev_req line has a requirement that conflicts with the immutable requirement, we need to resolve it
+                # we KNOW that the immutable requirement will be of form package==version, so we can reliably pull out the version
+                # and check it against the specifier of the dev_req line.
+                immutable_version = next(iter(immutable_requirement.specifier)).version
+
+                if not package_requirement.specifier.contains(immutable_version):
+                    if should_log:
+                        logging.info(
+                            f"Dev req dependency {package_name}'s requirement specifier of {package_requirement} is not compatible with immutable requirement {immutable_requirement}."
```

Comment:
Nit: can we run `black` on this file? Some of these lines look long

Reply:
I thought I did I'll re-run it 👍

Reply:
![image](https://github.com/user-attachments/assets/dcf68981-bff5-4b48-80a6-d1964ce2592b)

apparently not? I think I need to add newlines to longer strings myself.